packages/ext-taquito/src/tzktReadProvider.ts

```diff
@@ -0,0 +1,195 @@
+} from '@taquito/rpc';
+import {
+  BigMapQuery,
+  BlockIdentifier,
+  SaplingStateQuery,
+  TzReadProvider
+} from '@taquito/taquito';
+import {
+  accountsGetBalanceAtLevel,
+  accountsGetByAddress,
+  accountsGetCounter, bigMapsGetKey, Block, blocksGet,
+  blocksGetByHash,
+  blocksGetByLevel, blocksGetCount,
+  contractsGetEntrypoints, contractsGetRawStorage,
+  contractsGetStorageSchema, headGet, Protocol, protocolsGetByCycle,
+  protocolsGetCurrent
+} from '@tzkt/sdk-api';
+import {BigNumber} from 'bignumber.js';
+
+
+
+export class TzktReadProvider implements TzReadProvider {
+  constructor(private readProvider: TzReadProvider) {}
+
+  private readonly liveBlockFilters = {
+    level: {},
+    sort: {
+      desc: 'level',
+    },
+    limit: 120, select: {fields: ['hash', 'timestamp']}
+  };
+
+
+  private async _getBlockIdentifier(block: BlockIdentifier): Promise<number> {
+    if(typeof block === 'number') {
+      return block
+    }
+
+    if(block === 'head' || !block) {
+      const {data} = await blocksGetCount();
+      return data
+    }
+
+    if(String(block).includes('head~')) {
+      const n = Number(String(block).split('head~')[1])
+      const {data} = await blocksGetCount();
+
+      return  Number(data) - n
+    }
+
+    const {data} = await blocksGetByHash(block);
+    return data?.level || 0
+  }
+
+  async getEntrypoints(contract: string): Promise<EntrypointsResponse> {
+    const {data} = await contractsGetEntrypoints(contract);
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    return data;
+  }
+
+  async getScript(address: string): Promise<ScriptedContracts> {
+    const {data} = await contractsGetStorageSchema(address);
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    return data;
+  }
+
+  async getBalance(address: string, block: BlockIdentifier): Promise<BigNumber> {
+    const blockId = await this._getBlockIdentifier(block);
+    const {data} = await accountsGetBalanceAtLevel(address, blockId);
+    return new BigNumber(data);
+  }
+
+  async getDelegate(address: string): Promise<string | null> {
+    const {data} = await accountsGetByAddress(address);
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    return data?.delegate?.address
+  }
+
+  async getNextProtocol(): Promise<string> {
+    const {data} = await protocolsGetCurrent()
+    return data?.hash || ''
+  }
+
+  async getProtocolConstants(block: BlockIdentifier): Promise<{
+    time_between_blocks?: BigNumber[] | undefined;
+    minimal_block_delay?: BigNumber | undefined;
+    hard_gas_limit_per_operation: BigNumber;
+    hard_gas_limit_per_block: BigNumber;
+    hard_storage_limit_per_operation: BigNumber;
+    cost_per_byte: BigNumber;
+  }> {
+
+    let protocol: Protocol
+
+    if(block) {
+      const blockId = await this._getBlockIdentifier(block)
+      const blockByLevel = await blocksGetByLevel(blockId);
```

Comment:
if it's hash there is getBlockByHash call
If it's head you can query getBlocks with limit=1 sort desc by level